metric/precipitation_metrics.py

```diff
@@ -0,0 +1,135 @@
+        self.add_state("total_samples", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("total_pixels", default=torch.tensor(0), dist_reduce_fx="sum")
+        
+        # Add states for classification metrics
+        self.add_state("total_tp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("total_fp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("total_tn", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("total_fn", default=torch.tensor(0), dist_reduce_fx="sum")
+    
+    def update(self, preds, target):
+        """
+        Update the metric states with new predictions and targets.
+        
+        Args:
+            preds: Model predictions (normalized)
+            target: Ground truth (normalized)
+        """
+        # Move tensors to the same device as the metric states
+        preds = preds.to(self.device)
+        target = target.to(self.device)
+        
+        # Make sure preds and target have the same shape
+        if preds.shape != target.shape:
+            if len(preds.shape) < len(target.shape):
+                preds = preds.unsqueeze(0)
+            elif len(preds.shape) > len(target.shape):
+                preds = preds.squeeze()
+                # If squeezing made it too small, add dimension back
+                if len(preds.shape) < len(target.shape):
+                    preds = preds.unsqueeze(0)
+        
+        # Calculate MSE loss (normalized)
+        batch_size = target.size(0)
+        loss = torch.nn.functional.mse_loss(preds, target, reduction="sum") / batch_size
+        self.total_loss += loss
+        self.total_samples += 1
+        self.total_pixels += target.numel()
+        
+        # Denormalize if needed
+        if self.denormalize:
+            preds_denorm = preds * self.factor
+            target_denorm = target * self.factor
+            
+            # Calculate denormalized MSE loss
+            loss_denorm = torch.nn.functional.mse_loss(preds_denorm, target_denorm, reduction="sum") / batch_size
+            self.total_loss_denorm += loss_denorm
+            
+            # Convert to mm/h for classification metrics (multiply by 12 for 5min to hourly rate)
+            preds_hourly = preds_denorm * 12
+            target_hourly = target_denorm * 12
+            
+            # Apply threshold to get binary masks
+            preds_mask = preds_hourly > self.threshold
+            target_mask = target_hourly > self.threshold
+            
+            # Compute confusion matrix
+            confusion = target_mask.view(-1) * 2 + preds_mask.view(-1)
+            bincount = torch.bincount(confusion, minlength=4)
+            
+            # Update confusion matrix states
+            self.total_tn += bincount[0]
+            self.total_fp += bincount[1]
+            self.total_fn += bincount[2]
+            self.total_tp += bincount[3]
```

Comment:
These should also get updated when we not denormalize, right? Then, this should be unindented.